src/SpectatorView.Native/SpectatorView.Compositor/Compositor/VideoEncoder.h

```diff
@@ -38,33 +38,113 @@ class VideoEncoder
     void StopRecording();
 
     // Used for recording video from a background thread.
-    void QueueVideoFrame(byte* buffer, LONGLONG timestamp, LONGLONG duration);
+    class VideoInput;
+    std::unique_ptr<VideoInput> GetAvailableVideoFrame();
+    void QueueVideoFrame(std::unique_ptr<VideoInput> frame);
     void QueueAudioFrame(byte* buffer, int bufferSize, LONGLONG timestamp);
 
     // Do not call this from a background thread.
     void Update();
 
-private:
-    void WriteVideo(byte* buffer, LONGLONG timestamp, LONGLONG duration);
-    void WriteAudio(byte* buffer, int bufferSize, LONGLONG timestamp);
+    class VideoInputBase
+    {
+    public:
+        IMFMediaBuffer* mediaBuffer = nullptr;
+        LONGLONG timestamp = INVALID_TIMESTAMP;
```

Comment:
If we have concerns around exposing IMFMediaBuffer to the CompositorInterface, we could also add a custom typedef here. I don't know that we will ever support this compositor on mac, but we could have a platform specific type as needed.

Reply:
granted looking further below there are public constructors requiring references to other windows types (ID3D11Device, etc). I don't think there is a substantial difference to having the compositor interface know about media foundation types if it already knows about direct x types.

---
In reply to: [443729465](https://github.com/microsoft/MixedReality-SpectatorView/pull/395#discussion_r443729465) [](ancestors = 443729465)